### PR TITLE
Issue-4852: update segment split/merge metrics after rolling transaction is committed

### LINKED ISSUE #4852: Update segment split/merge metrics when rolling transaction commits
**Problem description**
Currently rolling transaction related scaling is done internally without updating segment split/merge metrics, which causes inconsistency between metrics and RestAPI(metadata)

**Problem location**
AbstractStreamMetadataStore.java

**Suggestions for an improvement**
After rolling transaction has been completed, retrieve the number of segment splits/merges and call Metrics API to update the corresponding metrics.

### PR DESCRIPTION
**Pull request by kevinhan88**
Signed-off-by: Kevin Han <kevinhan88@gmail.com>

**Change log description**  
After rolling transaction has been committed, update segment split/merge metrics with latest values.

**Purpose of the change**  
Fixes #4852 

**What the code does**  
Inside AbstractStreamMetadataStore, after completeRollingTxn, findNumSplitsMerges is called to get the latest values, followed by the call to update the metrics 

**How to verify it**  
Unit test should pass; metrics should be consistent with RestAPI in case of rolling transaction.

### CODE REVIEW

File: test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java [hunk 1]
```diff
@@ -229,4 +229,50 @@ public void testTransactionMetrics() throws Exception {
...
     }
+
+    @Test(timeout = 30000)
+    public void testRollingTxnMetrics() throws Exception {
+        String scaleRollingTxnScopeName = "scaleRollingTxnScope";
+        String scaleRollingTxnStreamName = "scaleRollingTxnStream";
+
+        controllerWrapper.getControllerService().createScope(scaleRollingTxnScopeName).get();
+        if (!controller.createStream(scaleRollingTxnScopeName, scaleRollingTxnStreamName, config).get()) {
+            log.error("Stream {} for scale testing already existed, exiting", scaleRollingTxnScopeName + "/" + scaleRollingTxnStreamName);
```
Comment (eolivelli) [inline test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java:240]: Why don't you fail the test?

File: test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java [hunk 2]
```diff
@@ -229,4 +229,50 @@ public void testTransactionMetrics() throws Exception {
...
+
+        //split to 3 segments
+        Map<Double, Double> keyRanges = new HashMap<>();
+        keyRanges.put(0.0, 0.25);
+        keyRanges.put(0.25, 0.75);
+        keyRanges.put(0.75, 1.0);
+
+        Stream scaleRollingTxnStream = new StreamImpl(scaleRollingTxnScopeName, scaleRollingTxnStreamName);
+        if (!controller.scaleStream(scaleRollingTxnStream, Collections.singletonList(0L), keyRanges, executor).getFuture().get()) {
+            log.error("Scale stream: splitting segment into three failed, exiting");
```
Comment (eolivelli) [inline test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java:261]: Same here

File: test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java [hunk 3]
```diff
@@ -229,4 +229,50 @@ public void testTransactionMetrics() throws Exception {
...
+        }
+
+        assertEquals(3, (long) MetricRegistryUtils.getGauge(MetricsNames.SEGMENTS_COUNT, streamTags(scaleRollingTxnScopeName, scaleRollingTxnStreamName)).value());
+        assertEquals(1, (long) MetricRegistryUtils.getGauge(MetricsNames.SEGMENTS_SPLITS, streamTags(scaleRollingTxnScopeName, scaleRollingTxnStreamName)).value());
+        assertEquals(0, (long) MetricRegistryUtils.getGauge(MetricsNames.SEGMENTS_MERGES, streamTags(scaleRollingTxnScopeName, scaleRollingTxnStreamName)).value());
+
+        transaction.flush();
+        transaction.commit();
+
+        Thread.sleep(1000);
```
Comment (eolivelli) [inline test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java:272]: This sleep may turn the test into a new flaky one.
You could wait for the conditions below to be verified

Comment (andreipaduroiu) [reply test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java:272]: +1. Please do not add arbitrary thread sleeps in tests

Comment (andreipaduroiu) [reply test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java:272]: Use `TestUtils.await` or `AssertExtensions.assertEventuallyEquals`

File: test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java (hunk 1 above)
Comment (andreipaduroiu) [reply test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java:240]: Fail the test.

Comment (kevinhan88) [reply test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java:240]: fixed

File: test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java (hunk 2 above)
Comment (kevinhan88) [reply test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java:261]: fixed

File: test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java (hunk 3 above)
Comment (kevinhan88) [reply test/integration/src/test/java/io/pravega/test/integration/StreamMetricsTest.java]: fixed, sorry for the overlooking, the extensions are being used already in the rest of the class.

File: controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java [hunk 4]
```diff
@@ -494,7 +494,13 @@ public OperationContext createContext(String scope, String name) {
     @Override
     public CompletableFuture<Void> completeRollingTxn(String scope, String name, Map<Long, Long> sealedActiveEpochSegments,
                                                       VersionedMetadata<CommittingTransactionsRecord> record, OperationContext context, Executor executor) {
-        return withCompletion(getStream(scope, name, context).completeRollingTxn(sealedActiveEpochSegments, record), executor);
+
+        CompletableFuture<Void> future = withCompletion(getStream(scope, name, context).completeRollingTxn(sealedActiveEpochSegments, record), executor);
+
+        future.thenCompose(result -> findNumSplitsMerges(scope, name, context, executor).thenAccept(simpleEntry ->
```
Comment (eolivelli) [inline controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java:500]: It looks like that this code is the same as in _scaleSegmentsSealed_
what about creating a common method ?

I am still new with this code, maybe @shrids can give more hints

Comment (kevinhan88) [reply controller/src/main/java/io/pravega/controller/store/stream/AbstractStreamMetadataStore.java:500]: No, scaleSegmentSealed updates activeSegment in addition to what rollingTxn does here.
If we create a common method, then we have to use if condition. In the end we may have one or two more lines of codes.

**Pull request merged**